pkg/tenant/projects.go

```diff
@@ -522,6 +568,155 @@ func (s *Server) ProjectUpdate(c *gin.Context) {
 	c.JSON(http.StatusOK, p.ToAPI())
 }
 
+// ProjectPatch applies a partial update to a project identified by the project ID in
+// the URL and the tenant ID in the request body.
+//
+// Route: /project/:projectID
+func (s *Server) ProjectPatch(c *gin.Context) {
+	var (
+		err              error
+		orgID, projectID ulid.ULID
+	)
+
+	// orgID is required to check ownership of the project
+	if orgID = orgIDFromContext(c); ulids.IsZero(orgID) {
+		return
+	}
+
+	// Parse the projectID from the URL
+	projectParam := c.Param("projectID")
+	if projectID, err = ulid.Parse(projectParam); err != nil {
+		sentry.Warn(c).Err(err).Str("projectID", projectParam).Msg("could not parse project id from URL")
+		c.JSON(http.StatusNotFound, api.ErrorResponse("project not found"))
+		return
+	}
+
+	// Patch the project, this method handles the rest of the responses and errors.
+	s.projectPatch(c, orgID, projectID)
+}
+
+// projectPatch is a handler for multiple patch project routes that applies a partial
+// update to a project. Only the provided fields are updated, and an error is returned
+// if any field does not exist or if no updates were made.
+func (s *Server) projectPatch(c *gin.Context, orgID, projectID ulid.ULID) {
+	var (
+		err     error
+		project *db.Project
+		patch   map[string]interface{}
+	)
+
+	// Parse the patch fields from the request body
+	if err = c.BindJSON(&patch); err != nil {
+		sentry.Warn(c).Err(err).Msg("could not parse project patch request")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse(api.ErrUnparsable))
+		return
+	}
+
+	if len(patch) == 0 {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("no fields provided for patch"))
+		return
+	}
+
+	// Verify that the project exists in the user's organization.
+	if err = db.VerifyOrg(c, orgID, projectID); err != nil {
+		if errors.Is(err, db.ErrNotFound) || errors.Is(err, db.ErrOrgNotVerified) {
+			c.JSON(http.StatusNotFound, api.ErrorResponse("project not found"))
+			return
+		}
+		sentry.Warn(c).Err(err).Msg("could not check verification")
+		c.JSON(http.StatusUnauthorized, api.ErrorResponse("could not verify organization"))
+		return
+	}
+
+	// Retrieve the project from the database.
+	if project, err = db.RetrieveProject(c, projectID); err != nil {
+		if errors.Is(err, db.ErrNotFound) {
+			c.JSON(http.StatusNotFound, api.ErrorResponse("project not found"))
+			return
+		}
+		sentry.Error(c).Err(err).Msg("could not retrieve project")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not lookup project"))
+		return
+	}
+
+	// Iterate over the patch fields and perform the updates.
+	for field, value := range patch {
+		var ok bool
+		switch field {
+		case "name":
+			var name string
+			if name, ok = value.(string); !ok {
+				c.JSON(http.StatusBadRequest, api.ErrorResponse(api.FieldTypeError(field, "string")))
+				return
+			}
+
+			if name == project.Name {
+				c.JSON(http.StatusBadRequest, api.ErrorResponse("project already has this name"))
+				return
+			}
```

Comment:
I don't think a 400 is necessary in this case. PATCH shouldn't preclude the PUT case (e.g. the front-end might send an entire data representation for PATCH and that's fine).

Reply:
I suppose if a non-editable field is passed in we can return a 400.